### PR TITLE
created fetch_tse_data

### PR DESCRIPTION
**Pull request by rafonseca**
Hello guys.
This is the script that fetches data from TSE website in order to create a list of brazilian politicians. I have also a small example notebook that uses this data.
As the script will be code reviewed, I thought the notebook version would be appreciated. So this is a direct export from a notebook. Is it ok?

### CODE REVIEW

**Comment by jtemporal** [comment]
Hi @rafonseca awesome contribution! I would recommend a few things here in order to start reviewing this PR =)

The `.py` generated from the notebook is awesome but it should be in the `develop/` directory alongside with the `.ipynb` file and the `.html` file. You can checkout [here](https://github.com/datasciencebr/serenata-de-amor/blob/master/CONTRIBUTING.md#best-practices) how to configure your Jupyter Notebook to generate all of these files while saving (if you haven't done so yet).

Also I would like to point out that inside the `src/` only goes the clean code, you can checkout other python scripts there if want ;)

So tl;dr for this PR. You'll need to do the following:
- [ ] move the `.py` to the `develop/` subdirectory
- [ ] commit both `.html` and `.ipynb` files in the `develop/` subdirectory
- [ ] create the clean `.py` script inside the `src/` subdirectory

**Comment by rafonseca** [comment]
Ok, I've got it.
so, should I close this PR and create another one?

**Comment by jtemporal** [comment]
No! you can do your changes here ;)
That's the workflow: you send things, we suggest some changes, you change it and it goes under review again. You are on the right track, great job!

**Comment by cuducos** [comment]
Awesome contribution @rafonseca, awesome feedback @jtemporal!

No need to close the PR. You can work on changes on this branch and new commits are automagically added to this PR when pushed to GitHub ; )

**Comment by rafonseca** [comment]
It is indeed magical

File: develop/2016-08-08-irio-translate-dataset.html [hunk 1]
```diff
@@ -8456,8 +8456,8 @@
 }
 /* Flexible box model classes */
 /* Taken from Alex Russell http://infrequently.org/2009/08/css-3-progress/ */
-/* This file is a compatibility layer.  It allows the usage of flexible box 
-model layouts across multiple browsers, including older browsers.  The newest,
+/* This file is a compatability layer.  It allows the usage of flexible box
```
Comment (guizero) [inline develop/2016-08-08-irio-translate-dataset.html]: `compatibility` estava certo.

File: develop/2016-08-08-irio-translate-dataset.html [hunk 2]
```diff
@@ -8456,8 +8456,8 @@
 }
 /* Flexible box model classes */
 /* Taken from Alex Russell http://infrequently.org/2009/08/css-3-progress/ */
-/* This file is a compatibility layer.  It allows the usage of flexible box 
-model layouts across multiple browsers, including older browsers.  The newest,
+/* This file is a compatability layer.  It allows the usage of flexible box 
+model layouts accross multiple browsers, including older browsers.  The newest,
```
Comment (guizero) [inline develop/2016-08-08-irio-translate-dataset.html]: `across` estava certo.

File: develop/2016-08-08-irio-translate-dataset.html [hunk 3]
```diff
@@ -11357,7 +11357,7 @@
  * Author: Jupyter Development Team
  */
 /** WARNING IF YOU ARE EDITTING THIS FILE, if this is a .css file, It has a lot
- * of chance of being generated from the ../less/[samename].less file, you can
+ * of chance of beeing generated from the ../less/[samename].less file, you can
```
Comment (guizero) [inline develop/2016-08-08-irio-translate-dataset.html]: `being` estava certo

**Comment by jtemporal** [comment]
Hi again, thank you for contribution @rafonseca and thank you @guizero for the review.

Actually irio's notebook shouldn't change.
- first because they weren't changes due to incorrect logic or analysis or any improvement on that aspect;
- and second because we try not to change notebooks since ideally the have historical reasons for being that way, by historical reasons I mean: in some point in the future we will be able to get the same results as the ones in the notebook if we replicate the same conditions such as python version, packages version and data version.

@rafonseca would you mind changing it back? that would also correct @guizero comments ;)

**Comment by rafonseca** [comment]
Hi @jtemporal  and @guizero,
sorry for the mess. Probably, I opened an older version of this notebook and accidentaly commited it. I will change it back as soon as possible. Actually, I need to figure out how to do this. I think I need help there, hehe..

**Comment by cuducos** [comment]
> I think I need help there, hehe...

Quick and dirty way: find Irio's notebook here on GitHub, click _Raw_, download and replace your local copy with it. Add a commit saying something like _Reverting Irio's notebook original version_. 

Classy way: play with `git revert` on the commits that changed these files, skip these files, `git commit` and `git rebase` to organize new commits if needed ; )

**Comment by rafonseca** [comment]
Thanks for the tip @cuducos . Finally, I did a checkout on that file using the SHA. Not so dirty, not so classy. Inevitably, I will make further errors, so I will have the opportunity to try the classy way :)

File: src/fetch_tse_data.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+import numpy as np
+import os
+import urllib
+import zipfile
+import glob
+
+
+
+FILENAME_PREFIX='consulta_cand_'
+TEMP_PATH = '../data/tse_temp'
```
Comment (cuducos) [inline src/fetch_tse_data.py]: This might fail in Windows (there it's `\` not `/`). We recommend using `os.path.join` to make file and directory paths that would work in both worlds.

File: src/fetch_tse_data.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+TSE_CANDIDATES_URL='http://agencia.tse.jus.br/estatistica/sead/odsele/consulta_cand/'
+OUTPUT_DATASET_PATH = '../data/2017-03-31-tse-candidates.xz'
```
Comment (cuducos) [inline src/fetch_tse_data.py]: Same as line 12.

File: src/fetch_tse_data.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+os.makedirs(TEMP_PATH)
```
Comment (cuducos) [inline src/fetch_tse_data.py]: We could use `tempfile` to manage temporary directories, couldn't we?

File: src/fetch_tse_data.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+# setting year range from 2004 to 2016. this will be modified further to 'from 1994 to 2016'
```
Comment (cuducos) [inline src/fetch_tse_data.py]: Why do we generate years from 2004 to 2016 and, later, transform it? Wouldn't it worth it to start from 1994 right now?

File: src/fetch_tse_data.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+year_list=[str(year) for year in (range(2004,2017,2))]
```
Comment (cuducos) [inline src/fetch_tse_data.py]: Can we use `date.today().year` instead of a hardcoded `2017`?

File: src/fetch_tse_data.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+# Download files
+for year in year_list:
+    filename=FILENAME_PREFIX+year+'.zip'
```
Comment (cuducos) [inline src/fetch_tse_data.py]: Using `format` is recommended when concatenating more than 2 strings: `'{}{}.zip'.format(FILENAME_PREFIX, year)`.

File: src/fetch_tse_data.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    file_url=TSE_CANDIDATES_URL+filename
+    output_file=os.path.join(TEMP_PATH,filename)
```
Comment (cuducos) [inline src/fetch_tse_data.py]: Some Python best practices are to use spaces between the object being defined and its value, and also after comas (e.g. `output_file = os.path.join(TEMP_PATH, filename)` instead of `output_file=os.path.join(TEMP_PATH,filename)`). Check [PEP8](http://pep8.readthedocs.io/) or [`prospector`](https://pypi.python.org/pypi/prospector/) if you are interested in this hints on code quality ; )

File: src/fetch_tse_data.py [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+    filename=FILENAME_PREFIX+year+'.zip'
+    filepath=os.path.join(TEMP_PATH,filename)
+    zip_ref = zipfile.ZipFile(filepath, 'r')
+    zip_ref.extractall(TEMP_PATH)
+    zip_ref.close()
+
+
+# ### Adding the headers
+# The following headers were extracted from LEIAME.pdf in consulta_cand_2016.zip.
+header_consulta_cand_till2010=[
```
Comment (cuducos) [inline src/fetch_tse_data.py]: Considering this headers section:

- [ ] We don't need the `#` with no comments at the end of the lines
- [ ]  Also I don't think we need the headers you skip with the `#`
- [ ] And we need to translate the headers to English ([example](https://github.com/datasciencebr/serenata-toolbox/blob/master/serenata_toolbox/ceap_dataset.py#L51-L151))

File: src/fetch_tse_data.py [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+        elif ('2012' in file_i):
+            cand_df_i=(pd.read_csv('./'+file_i,sep=';',header=None,dtype=str,names=header_consulta_cand_at2012,encoding='iso-8859-1'))
+        else:
+            cand_df_i=(pd.read_csv('./'+file_i,sep=';',header=None,dtype=str,names=header_consulta_cand_till2010,encoding='iso-8859-1'))
+        cand_df=cand_df.append(cand_df_i[sel_columns])
+
+cand_df.index=cand_df.reset_index().index # this index contains no useful information
+
+# Exporting data
+cand_df.to_csv(OUTPUT_DATASET_PATH,encoding='iso-8859-1',compression='xz',header=True,index=False)
```
Comment (cuducos) [inline src/fetch_tse_data.py]: Can we use `utf-8`? (same for lines 240-244).

File: src/fetch_tse_data.py [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+###Concatenate all files in one pandas dataframe
+cand_df=pd.DataFrame()
+for year in year_list:
+    filesname=FILENAME_PREFIX+year+'*.txt'
+    filespath=os.path.join(TEMP_PATH,filesname)
+    files_of_the_year= sorted(glob.glob(filespath))
+    for file_i in files_of_the_year:
+        # the following cases do not take into account next elections. hopefully, TSE will add headers to the files
+        if ('2014' in file_i) or ('2016' in file_i):
+            cand_df_i=(pd.read_csv('./'+file_i,sep=';',header=None,dtype=str,names=header_consulta_cand_from2014,encoding='iso-8859-1'))
```
Comment (cuducos) [inline src/fetch_tse_data.py]: Is it recommended to use `np.str` instead of `str`, @jtemporal?

**Comment by rafonseca** [comment]
Hi @cuducos. Thanks for the feedback. 
I completely agree that the fetch notebook should not be included. I thought it could be useful for the reviewing process. But for the data, it is still not clear to me what dataset we want to keep:
-This script merges some data from TSE selecting some columns. This is dataset number 1 and we have it.
-The notebook 'use-tse-data' use dataset number 1 and produces dataset number 2, which is not saved in data/ (hell, it is a notebook!).
dataset2 is a nice readable list of all elected politicians as requested in the issue #196, but dataset1 has indeed more information...
Again, thanks for the review. I will work on the code soon.

File: src/fetch_tse_data.py (hunk 7 above)
Comment (rafonseca) [inline src/fetch_tse_data.py]: Well, the TSE informs that the data from 1994 to 2002 is not consistent, and they are working on this. I can try to get those data, but ...

File: src/fetch_tse_data.py (hunk 8 above)
Comment (rafonseca) [inline src/fetch_tse_data.py]: Yes, we can. It is more elegant, but the headers are hardcoded and there is a high probability that next election dataset will have different headers from all these...

File: src/fetch_tse_data.py (hunk 10 above)
Comment (rafonseca) [inline src/fetch_tse_data.py]: Very reasonable!

File: src/fetch_tse_data.py (hunk 12 above)
Comment (rafonseca) [inline src/fetch_tse_data.py]: Probably yes for exporting and no for importing. I will check again

File: src/fetch_tse_data.py (hunk 7 above)
Comment (cuducos) [inline src/fetch_tse_data.py]: Wow… no, no… now I got it. Just leave code as it is, but maybe clarify it in comments…

File: src/fetch_tse_data.py (hunk 8 above)
Comment (cuducos) [inline src/fetch_tse_data.py]: Damn… that's not good news. That's something that might be helpful to have documented (in comments) to future-me and future-you ; )

File: src/fetch_tse_data.py (hunk 12 above)
Comment (cuducos) [inline src/fetch_tse_data.py]: Yep, sure thing. My bad. The idea as to use UTF-8 for exporting, sorry ; )

**Comment by cuducos** [comment]
Hum… now I got it.

I tend to agree that creating only the first dataset, and the notebook that demonstrated how to get from it to a list of elected politicians is good enough. 

@jtemporal and @Irio — you're more familiar than I am with data science — what do you think about it?

File: src/fetch_tse_data.py (hunk 13 above)
Comment (jtemporal) [inline src/fetch_tse_data.py]: Yes, indeed it is, in a few words: `<numpy str objects>` are different from `<python str objects>`. To keep the pattern used in the whole project we should stick to using `np.str` =)

**Comment by jtemporal** [comment]
I personally would go to having the first dataset on the S3 and keep the notebook that shows " how to get from it to a list of elected politicians".
We usually have this: A large, structured, translated and ready to work dataset from which we can retrieve information and/or intelligence by filtering or applying some coding on top of it.
Maybe @Irio has some other point of view to share with us.

**Comment by Irio** [comment]
@cuducos There's [a precedent](https://github.com/datasciencebr/serenata-de-amor/blob/fe9c5602c5ff496f8197777274b3c576faee1472/develop/2016-08-08-irio-translate-dataset.ipynb) for merging a notebook directly and only related to a `src` script, so I'm ok with keeping doing so.

**Comment by cuducos** [comment]
@Irio AFAIK the notebook you mention **does not** write to `data/` — it's a study about how to implement a script that keeps the data directory intact. If @rafonseca's notebook doesn't write new data I'm fine.

**Comment by Irio** [comment]
@rafonseca Do you need help for making these requested changes?

File: src/fetch_tse_data.py (hunk 13 above)
Comment (rafonseca) [inline src/fetch_tse_data.py]: Actually, I used "unicode" in python 2.7. In python 3, the standard str replaced unicode. We can use np.str as long as it deals with unicode characters. Does it?

**Comment by rafonseca** [comment]
Hi @Irio.
Indeed, I have been quite busy last weeks and I will probably be busy for the next two weeks. Shame on me. I will try to find some gap to work on this.
I will certainly need help for the headers translation.
There is one last thing which is not clear for me. As you have noticed, I created the script from the notebook and there is important information on the notebook that will help to maintain this script in the future, like those infos @cuducos asked to comment somewhere. So, we are going to keep also the original notebook or not? Of course, we would remove the file writing in '../data'

**Comment by rafonseca** [comment]
Hello there,
I did almost all modifications @cuducos  suggested. It is still missing the translation. Finally, I removed the scratch notebook and moved the important information to the script comments. I will try to answer more quickly for the next reviews. This PR is becoming long, hehe.
Thanks for the suggestions again

File: src/fetch_tse_data.py [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+import urllib
+import zipfile
+import glob
+
+from tempfile import mkdtemp
+TEMP_PATH= mkdtemp()
+
+FILENAME_PREFIX= 'consulta_cand_'
+TSE_CANDIDATES_URL= 'http://agencia.tse.jus.br/estatistica/sead/odsele/consulta_cand/'
+OUTPUT_DATASET_PATH= os.path.join(os.pardir,'data','2017-03-31-tse-candidates.xz')
```
Comment (cuducos) [inline src/fetch_tse_data.py]: It's not a good idea to have this date hardcoded here. Something like this might be useful:

```console
>>> from datetime import date
>>> today = date.today()
>>> today.strftime('%Y-%m-%d')
'2017-05-10'
```

**Comment by cuducos** [comment]
>  It is still missing the translation.

If the last link got lost, [this might me helpful](https://github.com/datasciencebr/serenata-toolbox/blob/master/serenata_toolbox/chamber_of_deputies/chamber_of_deputies_dataset.py#L51-L153) ; )

> I will try to answer more quickly for the next reviews.

No worries — worst case scenario if we need this data ASAP we merge your PR to a secondary branch and work on some tweaks before bringing your code to the master branch…

> This PR is becoming long

No problem at all! This is valuable data <3

**Comment by cuducos** [comment]
🎉 Many thanks @rafonseca ; )

**Comment by rafonseca** [comment]
yuhuu!!
Thank you all. Hope it is the first PR of many!
